File: frontend/src/services/accountService.js
import Account from '../models/Account';
import httpClient from './httpClient';

/**
 * Account Service
 * Handles account-related operations through the Golang Gin backend API
 */
const accountService = {
  /**
   * Get all accounts
   * @returns {Promise<Array>} - List of accounts
   */
  getAll: async () => {
    try {
      // Gin backend endpoint for accounts
      const data = await httpClient.backend.get('/accounts');
      return data.map(account => Account.fromJSON(account));
    } catch (error) {
      console.error('Error fetching accounts:', error);
      throw error;
    }
  },
  
  /**
   * Get account by ID
   * @param {string} id - Account ID
   * @returns {Promise<Object>} - Account data
   */
  getById: async (id) => {
    try {
      // Gin backend endpoint for single account
      const data = await httpClient.backend.get(`/accounts/${id}`);
      return Account.fromJSON(data);
    } catch (error) {
      console.error(`Error fetching account ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Create new account
   * @param {Object} accountData - Account data
   * @returns {Promise<Object>} - Created account
   */
  create: async (accountData) => {
    try {
      const data = await httpClient.backend.post('/accounts', accountData);
      return Account.fromJSON(data);
    } catch (error) {
      console.error('Error creating account:', error);
      throw error;
    }
  },
  
  /**
   * Update account
   * @param {string} id - Account ID
   * @param {Object} accountData - Updated account data
   * @returns {Promise<Object>} - Updated account
   */
  update: async (id, accountData) => {
    try {
      const data = await httpClient.backend.put(`/accounts/${id}`, accountData);
      return Account.fromJSON(data);
    } catch (error) {
      console.error(`Error updating account ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Delete account
   * @param {string} id - Account ID
   * @returns {Promise<Object>} - Response data
   */
  delete: async (id) => {
    try {
      return await httpClient.backend.delete(`/accounts/${id}`);
    } catch (error) {
      console.error(`Error deleting account ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Get account types
   * @returns {Promise<Array>} - List of account types
   */
  getAccountTypes: async () => {
    try {
      return await httpClient.backend.get('/accounts/types');
    } catch (error) {
      console.error('Error fetching account types:', error);
      throw error;
    }
  },
  
  /**
   * Get account summary
   * @returns {Promise<Object>} - Summary data
   */
  getSummary: async () => {
    try {
      // Gin backend endpoint for account summary
      return await httpClient.backend.get('/accounts/summary');
    } catch (error) {
      console.error('Error fetching account summary:', error);
      throw error;
    }
  }
};

export default accountService;

File: backend/internal/domain/services/account_service.go
package services

import (
	"github.com/quocdaijr/finance-management-backend/internal/domain/models"
	"github.com/quocdaijr/finance-management-backend/internal/repository"
)

// AccountService handles business logic for accounts
type AccountService struct {
	accountRepo *repository.AccountRepository
}

// NewAccountService creates a new account service
func NewAccountService(accountRepo *repository.AccountRepository) *AccountService {
	return &AccountService{
		accountRepo: accountRepo,
	}
}

// Create creates a new account
func (s *AccountService) Create(userID uint, req *models.AccountRequest) (*models.Account, error) {
	// Create account
	account := &models.Account{
		UserID:    userID,
		Name:      req.Name,
		Type:      req.Type,
		Balance:   req.Balance,
		Currency:  req.Currency,
		IsDefault: req.IsDefault,
	}

	// Save account
	if err := s.accountRepo.Create(account); err != nil {
		return nil, err
	}

	return account, nil
}

// GetByID gets an account by ID
func (s *AccountService) GetByID(id uint, userID uint) (*models.Account, error) {
	return s.accountRepo.GetByID(id, userID)
}

// GetAll gets all accounts for a user
func (s *AccountService) GetAll(userID uint) ([]models.Account, error) {
	return s.accountRepo.GetAll(userID)
}

// Update updates an account
func (s *AccountService) Update(id uint, userID uint, req *models.AccountRequest) (*models.Account, error) {
	// Get existing account
	account, err := s.accountRepo.GetByID(id, userID)
	if err != nil {
		return nil, err
	}

	// Update account
	account.Name = req.Name
	account.Type = req.Type
	account.Balance = req.Balance
	account.Currency = req.Currency
	account.IsDefault = req.IsDefault

	// Save account
	if err := s.accountRepo.Update(account); err != nil {
		return nil, err
	}

	return account, nil
}

// Delete deletes an account
func (s *AccountService) Delete(id uint, userID uint) error {
	return s.accountRepo.Delete(id, userID)
}

// GetAccountTypes gets all account types
func (s *AccountService) GetAccountTypes() []models.AccountType {
	return s.accountRepo.GetAccountTypes()
}

// GetSummary gets a summary of accounts for a user
func (s *AccountService) GetSummary(userID uint) (*models.AccountSummary, error) {
	return s.accountRepo.GetSummary(userID)
}
